docs/decisions/0039-function-call-content.md

```diff
@@ -0,0 +1,435 @@
+  * The `FunctionResult.Function` property is not deserializable and should be marked with the [JsonIgnore] attribute.  
+    * A new constructor, ctr(object? value = null, IReadOnlyDictionary<string, object?>? metadata = null), needs to be added for deserialization. 
+    * The `FunctionResult.Function` property has to be nullable. It can be a breaking change? for the function filter users because the filters use `FunctionFilterContext` class that expose an instance of kernel function via the `Function` property.
+
+### Option 3.4 - FunctionResult: KernelContent
+Note: This option was suggested during a second round of review of this ADR.
+   
+This option suggests making the `FunctionResult` class a derivative of the `KernelContent` class:
+```csharp
+public class FunctionResult : KernelContent
+{
+    ....
+}
+```
+So, instead of having a separate `FunctionResultContent` class to represent the function result content, the `FunctionResult` class will inherit from the `KernelContent` class, becoming the content itself. As a result, the function result returned by the `KernelFunction.InvokeAsync` method can be directly added to the `ChatMessageContent.Items` collection:
+```csharp
+foreach (FunctionCallContent functionCall in functionCalls)
+{
+    FunctionResult result = await functionCall.InvokeAsync(kernel);
+
+    chatHistory.Add(new ChatMessageContent(AuthorRole.Tool, new ChatMessageContentItemCollection { result }));
+    // instead of
+    chatHistory.Add(new ChatMessageContent(AuthorRole.Tool, new ChatMessageContentItemCollection { new FunctionResultContent(functionCall, result) }));
+    
+    // of cause, the syntax can be simplified by having additional instance/extension methods
+    chatHistory.AddFunctionResultMessage(result); // Using the new AddFunctionResultMessage extension method of ChatHistory class
+}
+```
+
+Questions:
+- How to pass the original `FunctionCallContent` to connectors along with the function result. It's actually not clear atm whether it's needed or not. The current rationale is that some models might expect properties of the original function call, such as arguments, to be passed back to the LLM along with the function result. An argument can be made that the original function call can be found in the chat history by the connector if needed. However, a counterargument is that it may not always be possible because the chat history might be truncated to save tokens, reduce hallucination, etc.
+- How to pass function id to connector?
+- How to communicate exception to the connectors? It was proposed to add the `Exception` property the the `FunctionResult` class that will always be assigned by the `KernelFunction.InvokeAsync` method. However, this change will break C# function calling semantic, where the function should be executed if the contract is satisfied, or an exception should be thrown if the contract is not fulfilled.
```

Comment:
Based on the filter work that @dmytrostruk is driving, I _think_ we want to have the following principle...
- Always throw an error (to match standard C# behavior)
- If invoked within the context of a kernel (e.g., prompt template or function calling), have a filter that exposes that the error occurred.
- At this point, the developer should have two options:
  1. Do nothing, in this case, the error should bubble up so it can be caught elsewhere
  2. Handle the error so that the error reason can be passed back to the AI

If we did this, would we even _need_ the `Exception` property since the exception reason would just be put in the main result field?

I'm assuming the main challenge would be maintaining backwards compatibility where we automatically handle exceptions on behalf of the user. Let me think on this some more...

Reply:
> If we did this, would we even _need_ the `Exception` property since the exception reason would just be put in the main result field?

if we do the `exception -> error message` translation inside a function filter and return the error message from the filter in a main result field, we won't need the `Exception` property.

> 
> I'm assuming the main challenge would be maintaining backwards compatibility where we automatically handle exceptions on behalf of the user. Let me think on this some more...

Do/will we handle exceptions automatically on behalf of users? Users will have to opt in by registering function filters to swallow/translate exceptions into error messages, won't they?